src/error.rs

```diff
@@ -5,14 +5,21 @@ use thiserror::Error;
 
 use crate::{java::lang::Throwable, Global, Jvm, Local};
 
+/// Result returned by most Java operations that may contain a local reference
+/// to a thrown exception.
 pub type Result<'jvm, T> = result::Result<T, Error<Local<'jvm, Throwable>>>;
+
+/// Result returned by [`crate::Jvm::with()`] that will store any uncaught
+/// exception as a global reference.
```

Comment:
Question: If exceptions are "exceptional", is it worth the optimization of having local references? Would be simpler to just *always* convert the exception to a global reference, no?

Reply:
It's a good question. In my head, Java exceptions are as exceptional as `Result` is in Rust---not very exceptional. Folks use them to express expected conditions all the time and they're fairly optimized within the JVM. For example, the auth code needs to check for the expected case of a few different auth failure types. They're more rare than the happy case, but not infrequent. 

Loosely speaking, I think it would be like declaring a single `Result<T, Box<dyn Error + Send>>` alias in Rust and using it everywhere. It would avoid some typing and performs fine most of the time. 

---

Thinking more about it, we only materialize the exception object during `.catch()` or during `.execute()`. We could keep it as a local for `.catch()` and then leave the result type as a global.

I'm assuming you're asking from the perspective of simplifying the result type to remove that `'jvm` lifetime?

Reply:
I think I've convinced myself the Global cost is probably worth the simplicity for most cases.